### PR TITLE
Feature/1b0a3d7/javakeystore

### PR DESCRIPTION
**Pull request by louis878**

### CODE REVIEW

File: core/java/iesi-core/src/main/java/io/metadew/iesi/common/crypto/FrameworkCrypto.java [hunk 1]
```diff
@@ -19,15 +24,34 @@
...
+                String alias = Configuration.getInstance().getMandatoryProperty("iesi.security.encryption.alias").toString();
+                String keyJKS = new JavaKeystore().loadKey(password, keystoreLocation, alias);
+                this.aes = new AESEncryptBasic(keyJKS);
+            } else if (Configuration.getInstance().getProperty("iesi.security.encryption.key").isPresent()) {
+                this.aes = new AESEncryptBasic(Configuration.getInstance().getMandatoryProperty("iesi.security.encryption.key").toString());
+            } else {
+                LOGGER.warn(MessageFormat.format("Warning: don't use hardcoded encrypted key", this.aes));
+                this.aes = new AESEncryptBasic("c7c1e47391154a6a");
+            }
+        } catch (Exception e) {
```
[the remark this answers is not in the document]
Comment (Levizar) [reply core/java/iesi-core/src/main/java/io/metadew/iesi/common/crypto/FrameworkCrypto.java]: As Exception were all coming from the same operation, it was done in the class throwing the error

File: core/java/iesi-core/src/main/java/io/metadew/iesi/common/crypto/FrameworkCrypto.java [hunk 2]
```diff
@@ -19,15 +24,34 @@
...
-        this.aes = new AESEncryptBasic("c7c1e47391154a6a");
+
+    public FrameworkCrypto() {
+        try {
+            if (Configuration.getInstance().getProperty("iesi.security.encryption.alias").isPresent()) {
+                Console console = System.console();
+                char[] password = console.readPassword("Enter password");
+                String keystoreLocation = Configuration.getInstance().getMandatoryProperty("iesi.security.encryption.keystore-path").toString();
+                String alias = Configuration.getInstance().getMandatoryProperty("iesi.security.encryption.alias").toString();
+                String keyJKS = new JavaKeystore().loadKey(password, keystoreLocation, alias);
```
[the remark this answers is not in the document]
Comment (Levizar) [reply core/java/iesi-core/src/main/java/io/metadew/iesi/common/crypto/FrameworkCrypto.java]: Is done

File: core/java/iesi-core/src/main/java/io/metadew/iesi/common/crypto/JavaKeystore.java [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+import java.io.FileInputStream;
+import java.io.IOException;
+import java.security.KeyStore;
+import java.security.KeyStoreException;
+import java.security.NoSuchAlgorithmException;
+import java.security.UnrecoverableEntryException;
+import java.security.cert.CertificateException;
+import java.security.spec.InvalidKeySpecException;
+
+//keytool -importpass -storetype pkcs12 -alias mypass -keystore Desktop/myks.p12
```
[the remark this answers is not in the document]
Comment (Levizar) [reply core/java/iesi-core/src/main/java/io/metadew/iesi/common/crypto/JavaKeystore.java]: This is the CLI command to create the keystore manually

File: core/java/iesi-core/src/test/java/io/metadew/iesi/common/crypto/FrameworkCryptoTest.java [hunk 4]
```diff
@@ -0,0 +1,52 @@
...
+        String password = "foobar";
+        String currentDirectory = System.getProperty("user.dir");
+        String keystoreLocation = currentDirectory + "/src/test/resources/" + Configuration.getInstance().getMandatoryProperty("iesi.security.encryption.keystore-path").toString();
+
+        System.setIn(new ByteArrayInputStream(password.getBytes()));
+        Scanner scanner = new Scanner(System.in);
+        String userinput = scanner.nextLine();
+        String alias = "mypass";
+        String keyJKS = new JavaKeystore().loadKey(userinput.toCharArray(), keystoreLocation, alias);
+        assertEquals("c7c1e47391154a6a", keyJKS);
```
[the remark this answers is not in the document]
Comment (BerrevoetsRobbe) [reply core/java/iesi-core/src/test/java/io/metadew/iesi/common/crypto/FrameworkCryptoTest.java:26]: Also, use something different than hardcoded password in Java Keystore to be sure key is not defaulted to hardcoded key

File: core/java/iesi-core/src/test/java/io/metadew/iesi/common/crypto/FrameworkCryptoTest.java [hunk 5]
```diff
@@ -0,0 +1,52 @@
...
+        } catch (Exception e) {
+            assertEquals("java.io.IOException: Integrity check failed: java.security.UnrecoverableKeyException: Failed PKCS12 integrity checking", e.getMessage());
+        }
+    }
+
+    @Test
+    public void testKeyInConf() {
+        if (Configuration.getInstance().getProperty("iesi.security.encryption.key").isPresent()) {
+            assertEquals("c7c1e47391154a6a", Configuration.getInstance().getMandatoryProperty("iesi.security.encryption.key").toString());
+        }
```
Comment (BerrevoetsRobbe) [inline core/java/iesi-core/src/test/java/io/metadew/iesi/common/crypto/FrameworkCryptoTest.java:50]: Does this actually test the conf?
Also, use something different than hardcoded password in Java Keystore to be sure key is not defaulted to hardcoded key

File: core/java/iesi-core/src/test/java/io/metadew/iesi/common/crypto/JavaKeystoreTest.java [hunk 6]
```diff
@@ -0,0 +1,64 @@
...
+        String alias = "mypas";
+        KeyStore.SecretKeyEntry ske = (KeyStore.SecretKeyEntry) ks.getEntry(alias,
+                new KeyStore.PasswordProtection(password.toCharArray()));
+        SecretKeyFactory factory = SecretKeyFactory.getInstance("PBE");
+        try {
+            factory.getKeySpec(
+                    ske.getSecretKey(), PBEKeySpec.class);
+        } catch (NullPointerException e) {
+            assertEquals(null, e.getMessage());
+        }
```
Comment (BerrevoetsRobbe) [inline core/java/iesi-core/src/test/java/io/metadew/iesi/common/crypto/JavaKeystoreTest.java:62]: use AssertJ unit test library

File: core/java/iesi-core/src/main/java/io/metadew/iesi/common/crypto/FrameworkCrypto.java [hunk 7]
```diff
@@ -40,8 +40,8 @@ public FrameworkCrypto() {
         } else if (Configuration.getInstance().getProperty("iesi.security.encryption.key").isPresent()) {
             this.aes = new AESEncryptBasic(Configuration.getInstance().getMandatoryProperty("iesi.security.encryption.key").toString());
         } else {
-            log.warn(MessageFormat.format("Warning: don't use hardcoded encrypted key", this.aes));
-            this.aes = new AESEncryptBasic("c7c1e47391154a6a");
+            log.warn(MessageFormat.format("iesi.encryption.key: no encryption key provided. IESI will use the hardcoded key", null));
```
Comment (BerrevoetsRobbe) [inline core/java/iesi-core/src/main/java/io/metadew/iesi/common/crypto/FrameworkCrypto.java:43]: remove MessageFormat all together

File: core/java/iesi-core/src/test/java/io/metadew/iesi/common/crypto/FrameworkCryptoTest.java [hunk 8]
```diff
@@ -39,14 +40,15 @@ public void testJavaKeystoreWrongPassword() {
...
         }
     }
 
     @Test
     public void testKeyInConf() {
-        if (Configuration.getInstance().getProperty("iesi.security.encryption.key").isPresent()) {
-            assertEquals("c7c1e47391154a6a", Configuration.getInstance().getMandatoryProperty("iesi.security.encryption.key").toString());
+        if (!Configuration.getInstance().getProperty("iesi.security.encryption.key").isPresent()) {
+            assertThat("y8c1e47391154a6c").isEqualTo(Configuration.getInstance().getMandatoryProperty("iesi.security.encryption.key").toString());
+            assertThat("y8c1e47391154a6b").isEqualTo(not(Configuration.getInstance().getMandatoryProperty("iesi.security.encryption.key").toString()));
```
Comment (BerrevoetsRobbe) [inline core/java/iesi-core/src/test/java/io/metadew/iesi/common/crypto/FrameworkCryptoTest.java]: A test should not contain an if statement. You should delete this if statement.
Also, the goal of the test is to verify the encryption key used by FrameworkCrypto, not the key stored in configuration.
You will probably have to mock/spy the Configuration singleton to manipulate this behaviour

File: core/java/iesi-core/src/test/java/io/metadew/iesi/common/crypto/JavaKeystoreTest.java [hunk 9]
```diff
@@ -58,7 +58,7 @@ public void testEncryptWrongAlias() throws Exception {
             factory.getKeySpec(
                     ske.getSecretKey(), PBEKeySpec.class);
         } catch (NullPointerException e) {
-            assertEquals(null, e.getMessage());
+            assertThat(e.getMessage()).isNull();
         }
     }
```
Comment (BerrevoetsRobbe) [inline core/java/iesi-core/src/test/java/io/metadew/iesi/common/crypto/JavaKeystoreTest.java]: There exists an AssertJ test method that allows for testing if a certain exception is thrown

File: core/java/iesi-core/src/test/java/io/metadew/iesi/common/configuration/ConfigurationTest.java [hunk 10]
```diff
@@ -21,6 +22,8 @@ void initTest() {
 
         PluginConfiguration pluginConfiguration = PluginConfiguration.getInstance();
         GuardConfiguration guardConfiguration = GuardConfiguration.getInstance();
+        FrameworkCrypto.getInstance();
+        System.out.println(FrameworkCrypto.getInstance());
```
Comment (Levizar) [inline core/java/iesi-core/src/test/java/io/metadew/iesi/common/configuration/ConfigurationTest.java:26]: Does it have to stay there ?

**Pull request merged**